store/tikv/latch/scheduler.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package latch
+
+import (
+	"sync"
+)
+
+const lockChanSize = 100
+
+// LatchesScheduler is used to schedule latches for transactions.
+type LatchesScheduler struct {
+	latches  *Latches
+	unlockCh chan *Lock
+	closed   bool
+	sync.RWMutex
+}
+
+// NewScheduler create the LatchesScheduler.
+func NewScheduler(size int) *LatchesScheduler {
+	latches := NewLatches(size)
+	unlockCh := make(chan *Lock, lockChanSize)
+	scheduler := &LatchesScheduler{
+		latches:  latches,
+		unlockCh: unlockCh,
+		closed:   false,
+	}
+	go scheduler.run()
+	return scheduler
+}
+
+func (scheduler *LatchesScheduler) run() {
+	for lock := range scheduler.unlockCh {
+		wakeupList := scheduler.latches.release(lock, lock.commitTS)
+		if len(wakeupList) > 0 {
+			scheduler.wakeup(wakeupList)
+		}
+	}
+}
+
+func (scheduler *LatchesScheduler) wakeup(wakeupList []*Lock) {
+	for _, lock := range wakeupList {
+		if scheduler.latches.acquire(lock) != acquireLocked {
+			lock.wg.Done()
+		}
+	}
+}
+
+// Close closes LatchesScheduler.
+func (scheduler *LatchesScheduler) Close() {
```

Comment:
Well, to make a right implementation for `Close`, you have to maintain a reference count for `Lock` and `Unlock`, don't close the channel before the reference count is zero.
That's more complex than it should be.